src/uTensor/ops/ActivationFncs_kernels.hpp

```diff
@@ -95,5 +136,119 @@ void relu6_k(Tensor& out, const Tensor& in) {
+  T mSum = 0;
+  const TensorShape& inShape = in->get_shape();
+  int outer_dim = inShape.num_dims() -1;
+  int depth = inShape[outer_dim];
+  int out_side_numelems = 1;
+  for(int i = 0; i < inShape.num_dims(); i++){
+    out_side_numelems *= (i == outer_dim) ? 1: inShape[i];
+  }
+
+  for (int i = 0; i < out_side_numelems; i++) {
+    // exp(x[i])/sum(exp(x[i])) == exp(x[i]+C)/sum(exp(x[i]+C))
+    T max = std::numeric_limits<T>::lowest();
+    for(int j = 0; j < depth; j++){
+      max = std::max(max, static_cast<T>(in(i, j)));
+    }
+
+    T mSum = 0;
+    for(int j = 0; j < depth; j++){
+      T tmp = exp((static_cast<T>(in(i,j)) - max) * beta);
+      mSum += tmp;
+      in(i,j) = tmp;
+    }
+    for(int j = 0; j < depth; j++){
+      in(i, j)  = static_cast<T>(in(i, j)) / mSum;
+    }
+  }
+}
+template <typename T>
+void softmax_k(Tensor& out, const Tensor& in, T beta=1) {
+  T tmp;
+  T mSum = 0;
+  const TensorShape& inShape = in->get_shape();
+  int outer_dim = inShape.num_dims() -1;
+  int depth = inShape[outer_dim];
+  int out_side_numelems = 1;
+  for(int i = 0; i < inShape.num_dims(); i++){
+    out_side_numelems *= (i == outer_dim) ? 1: inShape[i];
+  }
+
+  for (int i = 0; i < out_side_numelems; i++) {
+    // exp(x[i])/sum(exp(x[i])) == exp(x[i]+C)/sum(exp(x[i]+C))
+    T max = std::numeric_limits<T>::lowest();
+    for(int j = 0; j < depth; j++){
+      max = std::max(max, static_cast<T>(in(i, j)));
+    }
+
+    T mSum = 0;
+    for(int j = 0; j < depth; j++){
+      T tmp = exp((static_cast<T>(in(i,j)) - max) * beta);
+      mSum += tmp;
+      out(i,j) = tmp;
+    }
+    for(int j = 0; j < depth; j++){
+      out(i, j)  = static_cast<T>(out(i, j)) / mSum;
+    }
+  }
+
+}
+
+void sq_softmax_k(Tensor& out, const Tensor& in, int8_t beta=1);
+
+template <typename T>
+class inplace_sigmoid_k {
+  public:
+    void operator()(Tensor& t) const;
+};
+
+template <typename T>
+void inplace_sigmoid_k<T>::operator()(Tensor& t) const {
+  const T one = 1;
+  uint32_t t_size = t->get_shape().get_linear_size();
+  for (uint32_t i = 0; i < t_size; i++) {
+    const T tmp = one / (one + exp(- static_cast<T>(t(i))));
+    t(i) = tmp;
+  }
+}
+
+template <typename T>
+class sigmoid_k_impl {
+  public:
+    void operator()(Tensor& out, const Tensor& in) const;
+
+};
+
+template <typename T>
+void sigmoid_k_impl<T>::operator()(Tensor& out, const Tensor& in) const {
+  const T one = 1;
+  uint32_t t_size = in->get_shape().get_linear_size();
+  for (uint32_t i = 0; i < t_size; i++) {
+    const T tmp = one / (one + exp(- static_cast<T>(in(i))));
+    out(i) = tmp;
+  }
+}
+
+template <>
+void sigmoid_k_impl<int8_t>::operator()(Tensor& out, const Tensor& in) const;
+
+// Set defaults
+template <typename T>
+using sigmoid_k = sigmoid_k_impl<T>;
+
+template <typename T>
+using inplace_relu_k = inplace_relu_k_impl<T>;
+template <typename T>
+using relu_k = relu_k_impl<T>;
+template <typename T>
+using inplace_relu6_k = inplace_relu6_k_impl<T>();
```

Comment:
Why do you call the default constructor with `()` for `inplace_relu6_k` whereas without `()` for `inplace_relu_k`?
Is there a specific reason for this?

Reply:
hack

Reply:
So is `relu_k` a type or an object?
`inplace_relu6_k` is an object right?

Reply:
In both cases they are representing types here, you can think of it as a more useful method than a `#define`. Major issue is the verbosity in calling them, and how c++ does the initial preprocessing.

Reply:
How could `inplace_relu6_k_impl<T>()` be a type?
Can you elaborate it more?
And perhaps why you want to do it.

Reply:
lets take this chat offline